common/oauthTokenManager.go

```diff
@@ -713,15 +718,81 @@ func (credInfo *OAuthTokenInfo) GetNewTokenFromMSI(ctx context.Context) (*adal.T
+}
+
+// GetNewTokenFromMSI gets token from Azure Instance Metadata Service identity endpoint. It first checks if it is an Azure VM. Failing that case, it checks if the VM is registered with Azure Arc.
+// For details, please refer to https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/overview
+func (credInfo *OAuthTokenInfo) GetNewTokenFromMSI(ctx context.Context) (*adal.Token, error) {
+	// Try Arc VM
+	req, resp, err := credInfo.queryIMDS(MSIEndpointArcVM, Resource, IMDSAPIVersionArcVM, ctx)
 	if err != nil {
-		return nil, fmt.Errorf("please check whether MSI is enabled on this PC, to enable MSI please refer to https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/qs-configure-portal-windows-vm#enable-system-assigned-identity-on-an-existing-vm. (Error details: %v)", err)
+		// Try Azure VM
+		req, resp, err = credInfo.queryIMDS(MSIEndpointAzureVM, Resource, IMDSAPIVersionAzureVM, ctx)
+		if err != nil {
+			return nil, fmt.Errorf("please check whether MSI is enabled on this PC, to enable MSI please refer to https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/qs-configure-portal-windows-vm#enable-system-assigned-identity-on-an-existing-vm: %v", err)
+		}
+	} else {
+		challengeTokenPath := strings.Split(resp.Header["Www-Authenticate"][0], "=")[1]
+		// Open the file.
+		challengeTokenFile, fileErr := os.Open(challengeTokenPath)
+		if os.IsPermission(fileErr) {
+			if runtime.GOOS == "linux" {
+				return nil, fmt.Errorf("permission level inadequate to read Arc challenge token file %s. Make sure you are running AzCopy as a user who is a member of the \"himds\" group or is superuser.", challengeTokenPath)
+			} else if runtime.GOOS == "windows" {
+				return nil, fmt.Errorf("permission level inadequate to read Arc challenge token file %s. Make sure you are running AzCopy as a user who is a member of the \"local Administrators\" group or the \"Hybrid Agent Extension Applications\" group.", challengeTokenPath)
+			} else {
+				return nil, fmt.Errorf("error occurred while opening file %s in unsupported GOOS %s: %v", challengeTokenPath, runtime.GOOS, fileErr)
+			}
+		} else if fileErr != nil {
+			return nil, fmt.Errorf("error occurred while opening file %s: %v", challengeTokenPath, fileErr)
+		}
+
+		defer challengeTokenFile.Close()
+
+		// Create a new Reader for the file.
+		reader := bufio.NewReader(challengeTokenFile)
+		challengeToken, fileErr := reader.ReadString('\n')
+		if fileErr != nil && fileErr != io.EOF {
+			return nil, fmt.Errorf("error occurred while reading file %s: %v", challengeTokenPath, fileErr)
+		}
+
+		req.Header.Set("Authorization", "Basic "+challengeToken)
+
+		resp, err = msiTokenHTTPClient.Do(req)
+		if err != nil {
+			return nil, fmt.Errorf("failed to query token from Arc IMDS endpoint. Please report the issue to xxx@microsoft.com: %v", err)
```

Comment:
Reminder: Need to fix xxx@microsoft.com to a valid support channel id.

Reply:
Is this actually actionable for us? Perhaps the message should be about asking the user to verify their configuration first. They'll reach out via Github or Azure Support eventually if things don't work out.

Reply:
Yes, this need not be a support email but whatever we follow in other places, basically at this point it's not likely a user config error since we did get the proper challenge token file location in our first call to the ARC IMDS endpoint at L762 and then we could read the challenge token file at L772.
Since this is an ARC bug we want user to tell us that. Where shall we direct the user?

Reply:
@linuxsmiths typically we dont embed support details in production code if we can help it. What we do is return a specific/documented error code + message and then have separate support channels that we advertise separately from the code for customers to follow up on (_this could be github here for AzCopy customers OR our CSS routes for 1P services etc etc; But we don't need to make or embed the follow up route here in code - just the error code + actionable message_)

Reply:
@zezha-msft What would you suggest we put here instead of directing to the support mail? What should the actionable message be here?